### PR TITLE
Functioning Station module

### PR DESCRIPTION
**Pull request by prasadtalasila**
This PR does not incorporate the modified itinerary structure as incorporating the changes to a non functioning Station module would make any source of error ambiguous.

The Station Module is still in need of refactoring to improve code quality and semantics.

In the test for rapid processing of queries the Station module is unable to consume more than 10,000 queries in a second which implies that there is a significant bottleneck to be removed.

*** In the travis build rapid processing of queries fails at 1000 queries but on my system 10,000 queries are processed comfortably hence passing the test.

### CODE REVIEW

**Comment by SuhasHebbar** [comment]
From what I can see from the code climate issues they are mostly seem to be in the tests. The project modules seem to be clean. It seems credo only checks code climate issues for the project modules. The code climate issues are quite similar to that of what we had observed in credo. They will be fixed today.

**Comment by SuhasHebbar** [comment]
The code quality issue which you were referring to
```warning: redefining module MockCollector (current version defined in memory)
  deps/mox/lib/mox.ex:240

warning: redefining module MockRegister (current version defined in memory)
  deps/mox/lib/mox.ex:240```

occurs because the Registry and Collector mock defined in the station_fsm_test.exs and station_test.exs are of the same name. Since by default the mocks are not global they do not interfere with each other.

But in the case we had decided to have the mocks be global then there would have been a problem.
the solution to this would be to rename MockRegister in one of the test files to something else ( Maybe RegisterMock, but this seems quite ugly for a solution, hence I left it in the present state )

File: test/ts/station_fsm_test.exs [hunk 1]
```diff
@@ -0,0 +1,383 @@
+defmodule StationFsmTest do
+	@moduledoc"""
+	Test suite for the StationFsm module
+	"""
+
+	use ExUnit.Case, async: true
+	import Mox
+
+	setup_all do
+		Mox.defmock(MockCollector, for: TS.Collector)
```
Comment (prasadtalasila) [inline test/ts/station_fsm_test.exs]: These mocks can be renamed to `MockCollectorFsm and MockRegisterFsm

File: lib/ts/mock_station.ex [hunk 2]
```diff
@@ -0,0 +1,21 @@
+defmodule MockStation do
```
Comment (prasadtalasila) [inline lib/ts/mock_station.ex]: This file is being used in tests. Hence the right location for it is test/ts/mock_station.ex

**Comment by SuhasHebbar** [comment]
Also regarding the Station crashing. This occurs because when the test for the rapid processing of a stream of queries ends. The Station process still has 1000s of unprocessed queries in its message queue ( as it fails the test of processing 100,000 queries in 1 second).

To my understanding, at the end of the test the Station process has to be killed (```start_supervised({Station, [stationState, MockRegister, MockCollector]})```) and the session(for lack of a better term) of the Mocks also has to terminated for that test. If the Mocks are terminated first then when the station processes a query from the queue and the non existent mock is queried the Station crashes.

If one removes the negative test for the Station processing 100,000 queries the crash does not occur as then the station will not have any unprocessed queries in its queue.

**Comment by prasadtalasila** [comment]
@SuhasHebbar please provide links to code examples for suggested way of using mocks. I have the following questions about the elixir testing conventions:

1. What is the standard way of terminating a process in Elixir?
1. What is the standard way of terminating a process in a ExUnit test?
1. What is the standard way of terminating mocks (of Mox module) from a ExUnit test?

Please quote references from direct Elixir, ExUnit and Mox documentation if possible.

**Comment by SuhasHebbar** [comment]
If one moves mock_station.ex to test/ts/mock_station.ex then MockStation is no longer visible hence the tests fail.
Presently I'll have to see how one can add test/ts to the project PATH.

**Comment by prasadtalasila** [comment]
It is better to add the test/ts to project PATH. Please check for this solution.

**Comment by SuhasHebbar** [comment]
In the a barebone elixir process one normally does not have to manually terminate a process.
EVM should automatically terminate the process once the function which it is running terminates.

But if one wants to forcefully terminate a process [Process.exit/2](https://hexdocs.pm/elixir/Process.html#exit/2) can be used.

But in most cases one never has to interact with lower level or more barebone functions as most modules provide abstractions to terminate a process.

In case of a GenServer one can use [Genserver.stop/3](https://hexdocs.pm/elixir/GenServer.html#stop/3).
But in large scale distributed applications again one need not manually interact with the module specific termination functions as elixir gives the functionality of Supervision trees where the Supervisor can restart or stop child processes according to the given constraints.

According to https://hexdocs.pm/ex_unit/ExUnit.Callbacks.html#start_supervised/2
The advantage of starting a process under the test supervisor is that it is guaranteed to exit before the next test starts. Furthermore, because the child process is supervised, it will be restarted in case of crashes according to the :restart strategy in the child specification, even if stopped manually. Therefore, to guarantee a process started with start_supervised/2 terminates without restarts, see stop_supervised/1.

**Comment by prasadtalasila** [comment]
What if we can send an asynchronous message to Station module running inside a Genserver process to terminate itself? Such a message can be sent during the teardown stage of a test. Having such a message would also enable us to shut down the TS application in an orderly manner. Genserver seems to provide enough abstractions for orderly termination of a process. Let's stick to those.

Of course, the existing problem of the termination message sitting behind the burst of queries in mailbox still needs to be considered. Is there a way to fetch a different message from the mailbox queue on a priority basis?

Based on your explanation, I understand the termination of an Elixir process and a ExUnit process. As per your explanation on station crash, we can avoid the problem by terminating the station instead of the mocks. Is that understanding correct?

**Comment by SuhasHebbar** [comment]
It seems either priority of messages is either implemented internally in GenServer or GenServer.stop/1 or stop_supervised/1 work through a different mechanism as either functions successfully resolve the error.

**Comment by SuhasHebbar** [comment]
I found out how to change the project path based on the application environment from [this link](https://medium.com/perplexinomicon-of-philosodad/mock-modules-and-where-to-find-them-319ae74c088b).

The changes made are as follows
```def project do
    [
			app: :TransportScheduler,
			version: "0.1.0",
			elixir: "~> 1.5.3",
			test_coverage: [tool: ExCoveralls],
			preferred_cli_env: [coveralls: :test],
			source_url: "https://github.com/prasadtalasila/TransportScheduler",
			name: "TransportScheduler",
			start_permanent: Mix.env == :prod,
			elixirc_paths: elixirc_paths(Mix.env),
			deps: deps()
    ]
  end

	defp elixirc_paths(:test) do
		["lib", "test/mocks"]
	end

	defp elixirc_paths(_), do: ["lib"]

 ```

**Comment by prasadtalasila** [comment]
@SuhasHebbar Good that the PR has passed all the checks. Let's do a group code review for the PR.

In the mean time, please find out about the termination sequence and mailbox message reception using pattern matching. I am proposing a tentative code review session from 4PM to 6PM, 26-Jan-2018. Let's have @suryapratapmehra and @bgargi for the review session as well.

**Comment by prasadtalasila** [comment]
@SuhasHebbar 
I rebuilt the PR on travis. There is a unit test failure. Please see the [log](https://travis-ci.org/prasadtalasila/TransportScheduler/builds/333288699#L551).

**Comment by SuhasHebbar** [comment]
I've never faced that error on my system.
I've faced it once on Travis once or time.
I suspect that Travis does not assign the same system resources every time a build is runs as I cannot see any other reasons for non determinism for rapid query processing.

On another note, presently I am not able to access the local server. It would help as my system loaded with other applications running in the background would not give a good measure of if the application is running properly.

**Comment by SuhasHebbar** [comment]
I would also like to know your opinion on maintaining a baseline as pure GenServer as possible baseline implementation of the Station so that we have an idea of the overhead additions (or performance improvements) when we incorporate more features and modules into the Station.

**Comment by prasadtalasila** [comment]
@SuhasHebbar Here are clarifications to the points raised by you.

1. The server was in shutdown state. I restarted it. You can access it now.
1. You can add your suggestions on the improvements to the station module in issue #71 and in the [wiki page](https://github.com/prasadtalasila/TransportScheduler/wiki/TS-v1.0-Proposed-Architecture).
1. Its good to have a pure Genserver base line and then see the performance degradation due to the station code. The results here should correspond with the code profiling results (yet to be done) of the functions in the station module.

**Pull request merged**